File: android-app/app/src/main/java/com/audiobridge/Codecs.kt
package com.audiobridge

import org.concentus.OpusApplication
import org.concentus.OpusDecoder
import org.concentus.OpusEncoder
import kotlin.math.max
import kotlin.math.min

/** Mirrors pc-server/AudioBridge.Server/Core/Codecs.cs — kept symmetric so both sides agree on the
 *  wire format for a given transport without needing runtime codec negotiation. */

interface PcAudioCodec {
    /** Encode interleaved float PCM (-1..1) into wire bytes. */
    fun encode(interleaved: FloatArray, sampleCountPerChannel: Int): ByteArray

    /** Decode wire bytes into interleaved float PCM, returns sample count per channel. */
    fun decode(payload: ByteArray, outInterleaved: FloatArray): Int
}

interface MicAudioCodec {
    fun encode(pcm: ShortArray, sampleCount: Int): ByteArray
    fun decode(payload: ByteArray, outPcm: ShortArray): Int
}

/** Lossless passthrough — Wi-Fi and USB. */
class RawFloat32PcCodec : PcAudioCodec {
    override fun encode(interleaved: FloatArray, sampleCountPerChannel: Int): ByteArray {
        val bb = java.nio.ByteBuffer.allocate(interleaved.size * 4).order(java.nio.ByteOrder.LITTLE_ENDIAN)
        for (v in interleaved) bb.putFloat(v)
        return bb.array()
    }

    override fun decode(payload: ByteArray, outInterleaved: FloatArray): Int {
        val bb = java.nio.ByteBuffer.wrap(payload).order(java.nio.ByteOrder.LITTLE_ENDIAN)
        val count = payload.size / 4
        for (i in 0 until count) outInterleaved[i] = bb.float
        return count
    }
}

class RawPcm16MicCodec : MicAudioCodec {
    override fun encode(pcm: ShortArray, sampleCount: Int): ByteArray {
        val bb = java.nio.ByteBuffer.allocate(sampleCount * 2).order(java.nio.ByteOrder.LITTLE_ENDIAN)
        for (i in 0 until sampleCount) bb.putShort(pcm[i])
        return bb.array()
    }

    override fun decode(payload: ByteArray, outPcm: ShortArray): Int {
        val bb = java.nio.ByteBuffer.wrap(payload).order(java.nio.ByteOrder.LITTLE_ENDIAN)
        val count = payload.size / 2
        for (i in 0 until count) outPcm[i] = bb.short
        return count
    }
}

/** Bluetooth RFCOMM — fixed 48kHz/stereo, high bitrate (see PROTOCOL.md for the quality rationale). */
class OpusPcCodec(sampleRate: Int, private val channels: Int, bitrateBps: Int) : PcAudioCodec {
    private val encoder = OpusEncoder(sampleRate, channels, OpusApplication.OPUS_APPLICATION_AUDIO).apply {
        setBitrate(bitrateBps)
    }
    private val decoder = OpusDecoder(sampleRate, channels)
    private val scratch = ByteArray(8000)
    private val pcmScratch = ShortArray(sampleRate / 10 * channels)

    override fun encode(interleaved: FloatArray, sampleCountPerChannel: Int): ByteArray {
        val total = sampleCountPerChannel * channels
        for (i in 0 until total) {
            val v = min(1f, max(-1f, interleaved[i]))
            pcmScratch[i] = (v * Short.MAX_VALUE).toInt().toShort()
        }
        val len = encoder.encode(pcmScratch, 0, sampleCountPerChannel, scratch, 0, scratch.size)
        return scratch.copyOf(len)
    }

    override fun decode(payload: ByteArray, outInterleaved: FloatArray): Int {
        val frameSize = outInterleaved.size / channels
        val decoded = decoder.decode(payload, 0, payload.size, pcmScratch, 0, frameSize, false)
        val total = decoded * channels
        for (i in 0 until total) outInterleaved[i] = pcmScratch[i] / Short.MAX_VALUE.toFloat()
        return decoded
    }
}

class OpusMicCodec(sampleRate: Int, bitrateBps: Int) : MicAudioCodec {
    private val encoder = OpusEncoder(sampleRate, 1, OpusApplication.OPUS_APPLICATION_VOIP).apply {
        setBitrate(bitrateBps)
    }
    private val decoder = OpusDecoder(sampleRate, 1)
    private val scratch = ByteArray(4000)

    override fun encode(pcm: ShortArray, sampleCount: Int): ByteArray {
        val len = encoder.encode(pcm, 0, sampleCount, scratch, 0, scratch.size)
        return scratch.copyOf(len)
    }

    override fun decode(payload: ByteArray, outPcm: ShortArray): Int {
        return decoder.decode(payload, 0, payload.size, outPcm, 0, outPcm.size, false)
    }
}

File: pc-server/AudioBridge.Server/Core/Codecs.cs
using System;
using Concentus.Enums;
using Concentus.Structs;

namespace AudioBridge.Server.Core;

public enum PcAudioCodecKind : byte
{
    PcmFloat32 = 0x00,
    Opus = 0x01,
}

/// <summary>Encodes/decodes the PC->phone (system audio) stream.</summary>
public interface IPcAudioCodec
{
    PcAudioCodecKind Kind { get; }

    /// <summary>Encode one chunk of interleaved float PCM (range -1..1) into wire bytes.</summary>
    byte[] Encode(float[] interleaved, int sampleCountPerChannel);

    /// <summary>Decode wire bytes back into interleaved float PCM. Returns sample count per channel.</summary>
    int Decode(byte[] payload, float[] outInterleaved);
}

/// <summary>Encodes/decodes the phone->PC (mic) stream. Always PCM16 mono 16kHz canonical form.</summary>
public interface IMicAudioCodec
{
    byte[] Encode(short[] pcm, int sampleCount);
    int Decode(byte[] payload, short[] outPcm);
}

/// <summary>Lossless passthrough — used for Wi-Fi and USB where bandwidth is not a constraint.</summary>
public sealed class RawFloat32PcCodec : IPcAudioCodec
{
    public PcAudioCodecKind Kind => PcAudioCodecKind.PcmFloat32;

    public byte[] Encode(float[] interleaved, int sampleCountPerChannel)
    {
        var bytes = new byte[interleaved.Length * sizeof(float)];
        Buffer.BlockCopy(interleaved, 0, bytes, 0, bytes.Length);
        return bytes;
    }

    public int Decode(byte[] payload, float[] outInterleaved)
    {
        var count = payload.Length / sizeof(float);
        Buffer.BlockCopy(payload, 0, outInterleaved, 0, payload.Length);
        return count;
    }
}

/// <summary>Opus-compressed — used for Bluetooth RFCOMM, where raw stereo PCM would not reliably fit.</summary>
public sealed class OpusPcCodec : IPcAudioCodec
{
    private readonly OpusEncoder _encoder;
    private readonly OpusDecoder _decoder;
    private readonly int _channels;
    private readonly byte[] _scratch = new byte[8000];
    private readonly short[] _pcmScratch;

    public PcAudioCodecKind Kind => PcAudioCodecKind.Opus;

    public OpusPcCodec(int sampleRate, int channels, int bitrateBps)
    {
        _channels = channels;
        _encoder = new OpusEncoder(sampleRate, channels, OpusApplication.OPUS_APPLICATION_AUDIO)
        {
            Bitrate = bitrateBps,
        };
        _decoder = new OpusDecoder(sampleRate, channels);
        _pcmScratch = new short[sampleRate / 10 * channels]; // generous scratch (>100ms worth)
    }

    public byte[] Encode(float[] interleaved, int sampleCountPerChannel)
    {
        var total = sampleCountPerChannel * _channels;
        if (_pcmScratch.Length < total)
        {
            throw new InvalidOperationException("Opus scratch buffer too small for chunk");
        }
        for (var i = 0; i < total; i++)
        {
            var v = interleaved[i];
            if (v > 1f) v = 1f;
            if (v < -1f) v = -1f;
            _pcmScratch[i] = (short)(v * short.MaxValue);
        }
        var len = _encoder.Encode(_pcmScratch, 0, sampleCountPerChannel, _scratch, 0, _scratch.Length);
        var outBytes = new byte[len];
        Array.Copy(_scratch, outBytes, len);
        return outBytes;
    }

    public int Decode(byte[] payload, float[] outInterleaved)
    {
        var frameSize = outInterleaved.Length / _channels;
        var decoded = _decoder.Decode(payload, 0, payload.Length, _pcmScratch, 0, frameSize, false);
        var total = decoded * _channels;
        for (var i = 0; i < total; i++)
        {
            outInterleaved[i] = _pcmScratch[i] / (float)short.MaxValue;
        }
        return decoded;
    }
}

public sealed class RawPcm16MicCodec : IMicAudioCodec
{
    public byte[] Encode(short[] pcm, int sampleCount)
    {
        var bytes = new byte[sampleCount * sizeof(short)];
        Buffer.BlockCopy(pcm, 0, bytes, 0, bytes.Length);
        return bytes;
    }

    public int Decode(byte[] payload, short[] outPcm)
    {
        var count = payload.Length / sizeof(short);
        Buffer.BlockCopy(payload, 0, outPcm, 0, payload.Length);
        return count;
    }
}

public sealed class OpusMicCodec : IMicAudioCodec
{
    private readonly OpusEncoder _encoder;
    private readonly OpusDecoder _decoder;
    private readonly byte[] _scratch = new byte[4000];

    public OpusMicCodec(int sampleRate, int bitrateBps)
    {
        _encoder = new OpusEncoder(sampleRate, 1, OpusApplication.OPUS_APPLICATION_VOIP)
        {
            Bitrate = bitrateBps,
        };
        _decoder = new OpusDecoder(sampleRate, 1);
    }

    public byte[] Encode(short[] pcm, int sampleCount)
    {
        var len = _encoder.Encode(pcm, 0, sampleCount, _scratch, 0, _scratch.Length);
        var outBytes = new byte[len];
        Array.Copy(_scratch, outBytes, len);
        return outBytes;
    }

    public int Decode(byte[] payload, short[] outPcm)
    {
        return _decoder.Decode(payload, 0, payload.Length, outPcm, 0, outPcm.Length, false);
    }
}
